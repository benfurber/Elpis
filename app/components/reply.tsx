import React, { Component } from "react";
import { StyleSheet, Text, View } from "react-native";

import ActionSheet from "react-native-actionsheet";
import { TouchableOpacity } from "react-native-gesture-handler";

import { ActionSheetReply, Avatar, RichTextDisplay } from "components";
import { NavigationType, Reply as ReplyInterface } from "interfaces";
import { labels } from "labels";
import { colours, elements, layout } from "styles";
import { formatDate } from "utils";

interface Props {
  item: ReplyInterface;
  navigation: NavigationType;
}

interface State {
  deleting: boolean;
}

class Reply extends Component<Props, State> {
  ActionSheet = ActionSheet;

  state = {
    deleting: false,
  };

  onLongPress() {
    const { isAuthorCurrentUser } = this.props.item;

    if (isAuthorCurrentUser) {
      this.ActionSheet.show();
    }
  }

  render() {
    const { item, navigation } = this.props;
    const { author, edited, isAuthorCurrentUser, publishedAt } = item;
    const { deleting } = this.state;

    const opacity = deleting ? 0.5 : 1;

    return (
      <TouchableOpacity
        onLongPress={() => this.onLongPress()}
        disabled={!isAuthorCurrentUser}
      >
        <View style={[styles.commentContainer, { opacity }]}>
          <View style={styles.textContainer}>
            <RichTextDisplay item={item} navigation={navigation} />
          </View>
          <View style={styles.details}>
            <View style={styles.avatarContainer}>
              <Avatar
                avatarPath={author.avatarPath}
                navigation={navigation}
                size="large"
                userId={author.id}
              />
            </View>

            <View style={styles.authorDetails}>
              <Text style={styles.metaDetails}>
                <Text style={styles.name}>{`${author.name} `}</Text>
                {formatDate(publishedAt, false)}
                {edited && ` (${labels.editedReply})`}
              </Text>
            </View>
          </View>
        </View>
        <ActionSheetReply
          refProp={o => (this.ActionSheet = o)}
          navigation={navigation}
          item={item}
        />
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  authorDetails: {
    paddingLeft: layout.spacing,
  },
  avatarContainer: {
    ...elements.imageRoundLarge,
    alignItems: "center",
    borderColor: colours.pureWhite,
    borderWidth: 5,
    top: 15,
  },
  body: {
    paddingVertical: layout.spacing,
  },
  commentContainer: {
    backgroundColor: colours.pureWhite,
    borderRadius: layout.borderRadiusL,
    margin: layout.spacing,
    marginBottom: layout.spacingL,
    paddingHorizontal: layout.spacingL,
    paddingTop: layout.spacing,
  },
  details: {
    alignItems: "flex-end",
    flexDirection: "row",
  },
  metaDetails: {
    marginBottom: layout.spacing,
    ...elements.textDate,
  },
  name: {
    color: colours.navyBlueDark,
    fontWeight: "bold",
  },
  textContainer: {
    paddingTop: layout.spacing,
  },
});

export { Reply };
