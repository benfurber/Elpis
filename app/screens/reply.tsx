import React, { Component } from "react";
import { StyleSheet, View } from "react-native";
import { withMappedNavigationParams } from "react-navigation-props-mapper";

import {
  BackButton,
  BackgroundContainer,
  ButtonAddReply,
  ReplyList,
} from "components";
import { Comment as CommentType, NavigationType, Post } from "interfaces";
import { labels } from "labels";
import { colours } from "styles";
import { Analytics, getPostId, getTotalComments } from "utils";

import { Tabs } from "../components/post/tabs";

interface Props {
  backToText?: string;
  id: CommentType["id"];
  isFromTopic?: boolean;
  navigation: NavigationType;
  totalComments?: number;
  postId?: Post["id"];
}

interface State {
  postId: Post["id"] | null;
  totalComments: number | null;
}

class ReplyScreen extends Component<Props, State> {
  state = {
    postId: this.props.postId || null,
    totalComments: this.props.totalComments || null,
  };

  async componentDidMount() {
    const { id } = this.props;
    const { postId, totalComments } = this.state;

    Analytics.trackContent({
      contentId: id,
      contentType: "Replies",
    });

    if (totalComments === null) {
      const totalComments = await getTotalComments(id);
      totalComments && this.setState({ totalComments });
    }

    if (postId === null) {
      const postId = await getPostId(id);
      postId && this.setState({ postId });
    }
  }

  onPressComments() {
    const { backToText, isFromTopic, navigation } = this.props;
    const { postId } = this.state;

    if (isFromTopic) return navigation.pop();

    if (postId) {
      return navigation.push("Post", {
        backToText,
        id: postId,
        setDisplay: "comments",
      });
    }
  }

  onPressPost() {
    const { backToText, isFromTopic, navigation } = this.props;
    const { postId } = this.state;

    if (isFromTopic && backToText) return navigation.pop(2);
    if (isFromTopic) return navigation.popToTop();

    return navigation.push("Post", {
      backToText,
      id: postId,
      setDisplay: "body",
    });
  }

  render() {
    const { backToText, id, navigation } = this.props;
    const { totalComments } = this.state;
    const { toTopics } = labels.back;

    return (
      <BackgroundContainer>
        {backToText && (
          <BackButton
            navigation={navigation}
            onPress={() => navigation.popToTop()}
            text={backToText}
          />
        )}
        <Tabs
          onPressComments={() => this.onPressComments()}
          onPressPost={() => this.onPressPost()}
          display={"comments"}
          totalComments={totalComments || 0}
        />
        <View style={styles.container}>
          <BackButton
            navigation={navigation}
            text={toTopics}
            onPress={() => this.onPressComments()}
          />
          <ReplyList id={id} navigation={navigation} />
        </View>
        <ButtonAddReply commentId={id} navigation={navigation} />
      </BackgroundContainer>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: "stretch",
    backgroundColor: colours.whiteTransparent,
    flex: 1,
    width: "100%",
  },
});

const wrappedReplyScreen = withMappedNavigationParams()(ReplyScreen);
export {
  wrappedReplyScreen as ReplyScreen,
  ReplyScreen as UnwrappedReplyScreen,
};
