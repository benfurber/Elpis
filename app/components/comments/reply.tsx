import React, { Component } from "react";
import { StyleSheet, Text, View } from "react-native";

import { Avatar, Title } from "components";
import { Reply as ReplyInterface } from "interfaces";
import { colours, elements, layout } from "styles";
import { formatDate } from "utils";

interface Props {
  item: ReplyInterface;
}

class Reply extends Component<Props> {
  render() {
    const { item } = this.props;

    return (
      <View style={styles.commentContainer}>
        <View style={styles.details}>
          <Avatar avatarPath={item.author.avatarPath} size={"small"} />

          <View style={styles.authorDetails}>
            <Title text={item.author.name} small />
            <Text style={elements.textDate}>{formatDate(item.createdAt)}</Text>
          </View>
        </View>
        <View style={styles.textContainer}>
          <Text style={styles.body}>{item.content}</Text>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  authorDetails: {
    paddingLeft: layout.spacing,
  },
  body: {
    paddingVertical: layout.spacing,
  },
  commentContainer: {
    borderBottomColor: colours.pureWhite,
    borderBottomWidth: 3,
    margin: layout.spacing,
  },
  details: {
    alignItems: "center",
    flexDirection: "row",
    paddingVertical: layout.spacing,
  },
  textContainer: {
    paddingBottom: layout.spacingL,
  },
});

export { Reply };
