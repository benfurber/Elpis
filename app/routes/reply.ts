import { ReplyScreen } from "screens";

const path = "reply/:id";
const screen = ReplyScreen;

const Reply = {
  path,
  screen,
};

export { Reply };
