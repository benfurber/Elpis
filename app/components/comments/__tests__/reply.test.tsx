import "react-native";
import React from "react";
import { shallow } from "enzyme";

import { Reply } from "../reply";

describe("Comments<Reply>", () => {
  it("renders correctly", () => {
    const item = {
      author: {
        avatarPath: require("assets/images/profile-pic-annon.png"),
        name: "Kelli S"
      },
      body:
        "Meu pai também fez o mesmo. Logo depois que minha mãe pediu o divórcio porque sofria violência dentro de casa, ele pegou as malas, se mudou e sumiu no mundo. Nem faço mais questão de saber pra onde ele foi.",
      dateCreated: new Date("2019-01-02"),
      id: "21098"
    };

    const component = shallow(<Reply item={item} />);

    expect(component).toMatchSnapshot();
  });
});
