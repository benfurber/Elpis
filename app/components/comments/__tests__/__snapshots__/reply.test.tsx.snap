// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments<Reply> renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  delayLongPress={600}
  extraButtonProps={
    Object {
      "rippleColor": "transparent",
    }
  }
  onLongPress={[Function]}
>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "rgb(255, 255, 255)",
          "borderBottomWidth": 3,
          "margin": 10,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingVertical": 10,
        }
      }
    >
      <Avatar
        avatarPath={
          Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          }
        }
        size="small"
        userId="1235we"
      />
      <View
        style={
          Object {
            "paddingLeft": 10,
          }
        }
      >
        <Title
          small={true}
          text="Empodere Duas Mulheres"
        />
        <Text
          style={
            Object {
              "color": "rgb(51, 51, 51)",
              "fontStyle": "italic",
            }
          }
        >
          Há um ano
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingBottom": 20,
        }
      }
    >
      <RichTextDisplay
        item={
          Object {
            "author": Object {
              "avatarPath": Object {
                "testUri": "../../../app/assets/images/empower_two_women_logo.png",
              },
              "id": "1235we",
              "name": "Empodere Duas Mulheres",
              "posts": Array [],
            },
            "comment": Object {
              "author": Object {
                "avatarPath": Object {
                  "testUri": "../../../app/assets/images/empower_two_women_logo.png",
                },
                "id": "1235we",
                "name": "Empodere Duas Mulheres",
                "posts": Array [],
              },
              "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
              "edited": false,
              "id": "21097",
              "isAuthorCurrentUser": false,
              "publishedAt": 2019-01-01T00:00:00.000Z,
              "replies": Array [],
              "totalReplies": 0,
            },
            "content": "Nao.",
            "edited": false,
            "id": "21097",
            "isAuthorCurrentUser": false,
            "link": null,
            "publishedAt": 2019-01-01T00:00:00.000Z,
          }
        }
      />
    </View>
  </View>
  <ActionSheetReply
    item={
      Object {
        "author": Object {
          "avatarPath": Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          },
          "id": "1235we",
          "name": "Empodere Duas Mulheres",
          "posts": Array [],
        },
        "comment": Object {
          "author": Object {
            "avatarPath": Object {
              "testUri": "../../../app/assets/images/empower_two_women_logo.png",
            },
            "id": "1235we",
            "name": "Empodere Duas Mulheres",
            "posts": Array [],
          },
          "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
          "edited": false,
          "id": "21097",
          "isAuthorCurrentUser": false,
          "publishedAt": 2019-01-01T00:00:00.000Z,
          "replies": Array [],
          "totalReplies": 0,
        },
        "content": "Nao.",
        "edited": false,
        "id": "21097",
        "isAuthorCurrentUser": false,
        "link": null,
        "publishedAt": 2019-01-01T00:00:00.000Z,
      }
    }
    refProp={[Function]}
  />
</TouchableOpacity>
`;

exports[`Comments<Reply> renders correctly when edited 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  delayLongPress={600}
  extraButtonProps={
    Object {
      "rippleColor": "transparent",
    }
  }
  onLongPress={[Function]}
>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "rgb(255, 255, 255)",
          "borderBottomWidth": 3,
          "margin": 10,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingVertical": 10,
        }
      }
    >
      <Avatar
        avatarPath={
          Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          }
        }
        size="small"
        userId="1235we"
      />
      <View
        style={
          Object {
            "paddingLeft": 10,
          }
        }
      >
        <Title
          small={true}
          text="Empodere Duas Mulheres"
        />
        <Text
          style={
            Object {
              "color": "rgb(51, 51, 51)",
              "fontStyle": "italic",
            }
          }
        >
          Há um ano
           (Editada)
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingBottom": 20,
        }
      }
    >
      <RichTextDisplay
        item={
          Object {
            "author": Object {
              "avatarPath": Object {
                "testUri": "../../../app/assets/images/empower_two_women_logo.png",
              },
              "id": "1235we",
              "name": "Empodere Duas Mulheres",
              "posts": Array [],
            },
            "comment": Object {
              "author": Object {
                "avatarPath": Object {
                  "testUri": "../../../app/assets/images/empower_two_women_logo.png",
                },
                "id": "1235we",
                "name": "Empodere Duas Mulheres",
                "posts": Array [],
              },
              "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
              "edited": false,
              "id": "21097",
              "isAuthorCurrentUser": false,
              "publishedAt": 2019-01-01T00:00:00.000Z,
              "replies": Array [],
              "totalReplies": 0,
            },
            "content": "Nao.",
            "edited": true,
            "id": "21097",
            "isAuthorCurrentUser": false,
            "link": null,
            "publishedAt": 2019-01-01T00:00:00.000Z,
          }
        }
      />
    </View>
  </View>
  <ActionSheetReply
    item={
      Object {
        "author": Object {
          "avatarPath": Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          },
          "id": "1235we",
          "name": "Empodere Duas Mulheres",
          "posts": Array [],
        },
        "comment": Object {
          "author": Object {
            "avatarPath": Object {
              "testUri": "../../../app/assets/images/empower_two_women_logo.png",
            },
            "id": "1235we",
            "name": "Empodere Duas Mulheres",
            "posts": Array [],
          },
          "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
          "edited": false,
          "id": "21097",
          "isAuthorCurrentUser": false,
          "publishedAt": 2019-01-01T00:00:00.000Z,
          "replies": Array [],
          "totalReplies": 0,
        },
        "content": "Nao.",
        "edited": true,
        "id": "21097",
        "isAuthorCurrentUser": false,
        "link": null,
        "publishedAt": 2019-01-01T00:00:00.000Z,
      }
    }
    refProp={[Function]}
  />
</TouchableOpacity>
`;
